File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveNewCall.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveNewCall;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\NewExpression;

class ResolveNewCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveNewCalls");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof NewExpression;
    }
    
    function transform($expr) {
        $newExpr = $expr;
        for ($i = 0; $i < count($newExpr->args); $i++) {
            $newExpr->args[$i]->setExpectedType($newExpr->cls->decl->constructor_->parameters[$i]->type);
            $newExpr->args[$i] = $this->main->runPluginsOn($newExpr->args[$i]);
        }
        $expr->setActualType($newExpr->cls);
        return $expr;
    }
}

File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveNewCall.py
from onelang_core import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs

class ResolveNewCalls(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveNewCalls")
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.NewExpression)
    
    def transform(self, expr):
        new_expr = expr
        i = 0
        
        while i < len(new_expr.args):
            new_expr.args[i].set_expected_type(new_expr.cls_.decl.constructor_.parameters[i].type)
            new_expr.args[i] = self.main.run_plugins_on(new_expr.args[i])
            i = i + 1
        expr.set_actual_type(new_expr.cls_)
        return expr